### PR TITLE
New Resource: `azurerm_monitor_scheduled_query_rules_alert` & `azurerm_monitor_scheduled_query_rules_log`

### LINKED ISSUE #3951: Support alerts based on Log analytics queries
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
Create Alerts based on Log analytics queries.  as documented here https://docs.microsoft.com/en-us/azure/azure-monitor/platform/alerts-log
The corresponding API
https://docs.microsoft.com/en-us/rest/api/monitor/scheduledqueryrules/createorupdate
Currently only possible to create alerts based on Azure monitor metrics. 
https://www.terraform.io/docs/providers/azurerm/r/monitor_metric_alert.html
### New or Affected Resource(s)

* azurerm_monitor

### PR DESCRIPTION
**Pull request by mcdafydd**
Fixes #3951

This feature implements the [ScheduledQueryRules API](https://docs.microsoft.com/en-us/rest/api/monitor/scheduledqueryrules).  The new resources should allow users to automate creation of monitoring and alerting rules alongside infrastructure.

### CODE REVIEW

**Comment by mcdafydd** [comment]
I think I'm all good now and ready for review.

Thanks!

**Comment by anvarich** [comment]
@tombuildsstuff @katbyte would be great to have it 1.39.0 release ;)

**Comment by anvarich** [comment]
NY present maybe in this time 🙏

**Comment by SIGAN** [comment]
Please review this pull request - many people waiting for this change!

**Comment by anvarich** [comment]
@katbyte as i understood from tagging it will be merged in current 1.42 release cycle?

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+
+			"action_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"Alerting",
+					"LogToMetric",
+				}, false),
+			},
+			"authorized_resources": {
```
Comment (katbyte) [inline azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: Are these ids? if so 
```suggestion
			"authorized_resource_ids": {
```

might be more clear?

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
```
Comment (katbyte) [inline azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: Can we validate these?

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+			},
+			"azns_action": {
```
Comment (katbyte) [inline azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: azns sounds fairly opaque, is there a better name we could use here?

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go [hunk 4]
```diff
@@ -0,0 +1,658 @@
...
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"action_group": {
+							Type:     schema.TypeSet,
+							Required: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
```
Comment (katbyte) [inline azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: Could we validate this?

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"azns_action": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"action_group": {
```
Comment (katbyte) [inline azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: is this the name or id?

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go [hunk 6]
```diff
@@ -0,0 +1,658 @@
...
+					Schema: map[string]*schema.Schema{
+						"dimension": {
+							Type:     schema.TypeSet,
+							Required: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"name": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
```
Comment (katbyte) [inline azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: COuld we validate this with at least no empty stings?

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go [hunk 7]
```diff
@@ -0,0 +1,658 @@
...
+									"operator": {
+										Type:     schema.TypeString,
+										Required: true,
```
Comment (katbyte) [inline azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: If this has only one value, we could make this optional with a default?

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go [hunk 8]
```diff
@@ -0,0 +1,658 @@
...
+										ValidateFunc: validation.StringInSlice([]string{
+											"Include",
+										}, false),
+									},
+									"values": {
+										Type:     schema.TypeList,
+										Required: true,
+										Elem: &schema.Schema{
+											Type: schema.TypeString,
+										},
```
Comment (katbyte) [inline azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: COuld we validate this with at least no empty stings?

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go [hunk 9]
```diff
@@ -0,0 +1,658 @@
...
+									},
+								},
+							},
+						},
+						"metric_name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
```
Comment (katbyte) [inline azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: COuld we validate this with at least no empty stings?

File: azurerm/internal/services/monitor/tests/data_source_monitor_scheduled_query_rules_test.go [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceAzureRMMonitorScheduledQueryRules_logToMetricActionConfig(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
Comment (katbyte) [inline azurerm/internal/services/monitor/tests/data_source_monitor_scheduled_query_rules_test.go]: could we use
```suggestion
  name     = "acctestRG-monitor-%d"
```

for allr esource group names?

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go (hunk 1 above)
Comment (mcdafydd) [reply azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: Done!

File: azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go (hunk 3 above)
Comment (mcdafydd) [reply azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go]: I renamed this from azns_action to simply `action`

**Comment by mcdafydd** [comment]
I believe all the comments are addressed.  I requested a review but that didn't seem to remove the "waiting-response" label.

File: website/docs/r/monitor_scheduled_query_rules_alert.html.markdown [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  data_source_id = azurerm_application_insights.example.id
+  description    = "Scheduled query rule Alerting Action example"
+  enabled        = true
+  frequency      = 5
+  query          = "requests | where status_code >= 500 | summarize AggregatedValue = count() by bin(timestamp, 5m)"
+  query_type     = "ResultCount"
+  severity       = 1
+  time_window    = 30
+  trigger {
+    threshold_operator = "GreaterThan"
```
Comment (davidtom) [inline website/docs/r/monitor_scheduled_query_rules_alert.html.markdown]:
```suggestion
    operator = "GreaterThan"
```

Comment (davidtom) [reply website/docs/r/monitor_scheduled_query_rules_alert.html.markdown]: When using this resource I get an error that `threshold_operator` is not supported and that I'm missing the `operator` argument

Comment (mcdafydd) [reply website/docs/r/monitor_scheduled_query_rules_alert.html.markdown]: Thank you!  I pushed an update.

**Comment by mcdafydd** [comment]
> Hey @mcdafydd, I've got a doozy of a review for you but I appreciate the effort so far. My main concern is around maintainability and I've laid out the reasoning for that in a comment down below.

@mbfrahry, I just want to say thank you very much for the thoughtful review and the time to share your insights.  I'll start addressing these issues asap and try to push an update later tonight or tomorrow.  All your comments make sense to me.

I'm surprised by the crashing test.  I may need to update from master.

**Comment by mcdafydd** [comment]
I think we're getting closer.  I addressed a lot of the concerns you brought up, but still have a bit more to do on validations, helper functions, and tests.

For validations, I followed comments that appeared in the Azure portal and anything I could find in the documentation or when testing the APIs.

I'm thinking about creating a third resource called `azurerm_monitor_scheduled_query_rules_alert_metric`.  There seem to be three types of things you can do with the ScheduledQueryRules API today:

1. LogToMetricAction - near-realtime; stateful; limited set of supported tables (ie: Perf, Event, Heartbeat, Update) - creates an Azure Metric from a Log Analytics resource that can be alerted upon using a Metric Alert rule
2. AlertingAction with result count - stateless; runs action if the number of query results exceeds threshold
3. AlertingAction with metric trigger - stateless; requires time series aggregation and evaluates query results by metric column name and operator; runs action if the consecutive or total number of metric results exceeds threshold

The Azure portal gives this error when trying to create a metric trigger if the query isn't correct: `Search Query should contain 'AggregatedValue' and 'bin(TimeGenerated, [roundTo])' for Metric alert type`.  I was thinking that a new resource would be better here and allow us to validate the search query in the schema rather than in the CreateUpdate() function, but I'm open to suggestions.

**Comment by mbfrahry** [comment]
Glad to hear it @mcdafydd! 

I appreciate you taking the time to go through the validations. We try to keep it consistent with the UI and API but when it's too difficult to find we can just go with sane validation like no empty strings or integers greater than 0. The more precise the better though.

As for the third resource, we're dipping into even more complicated territory. We have too many potential routes to go down and no real best option. I'm hesitant to create a whole new resource if the only difference between the two is what the query can have in it. Are there more differences than just that?  It'd be hard to police from an import perspective as someone who doesn't understand this to the fullest could import incorrectly and then be forced to untangle how their resources should be saved in state. I think for this case just having an extra example in the documentation that shows how to use the trigger with the correct query will suffice. We could add additional validation around triggers and queries in CreateUpdate to confirm that the query has the things you're requesting.

**Comment by mcdafydd** [comment]
I'm going to do a once over on everything again this weekend, but I think it's close enough for re-review.  Thanks!

**Comment by mcdafydd** [comment]
I do see one issue that I'll try and address as soon as I can.  The LogToMetric tests and example don't reflect what's required to make that useful:

https://docs.microsoft.com/en-us/azure/azure-monitor/platform/alerts-metric-logs#resource-template-for-metric-alerts-for-logs

To use the LogToMetric API, there needs to be a second resource, an Azure Metric Alert, that alerts on the newly-created metric.  I'll update the tests and example to reflect this.

**Comment by mcdafydd** [comment]
Presuming this test passes, I'm ready for another review.

Thanks!

**Comment by mcdafydd** [comment]
Oops - I found one additional published limit - 100 resources maximum for a cross-resource query:

https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits#log-queries-and-language

**Comment by mcdafydd** [comment]
@tombuildsstuff @mbfrahry @katbyte 

Just checking in since this is my first new resource PR.  Is there anything else I should be doing to make sure it is mergeable to master?  I'll be patient, just want to be able to streamline any requirements if possible.

Thanks for all your work on 2.0!

File: azurerm/internal/services/monitor/tests/data_source_monitor_scheduled_query_rules_alert_test.go [hunk 12]
```diff
@@ -9,7 +9,7 @@ import (
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/acceptance"
 )
 
-func TestAccDataSourceAzureRMMonitorScheduledQueryRules_alertingAction(t *testing.T) {
+func TestAccDataSourceAzureRMMonitorScheduledQueryRulesAlertingAction(t *testing.T) {
```
Comment (mbfrahry) [inline azurerm/internal/services/monitor/tests/data_source_monitor_scheduled_query_rules_alert_test.go]: Let's keep the `_` that was removed in. It lets us quickly see and grab the prefix when we want to run this subset of tests and has some other minor benefits that warrant keeping it in

File: website/docs/r/monitor_scheduled_query_rules_alert.html.markdown [hunk 13]
```diff
@@ -25,16 +25,43 @@ resource "azurerm_application_insights" "example" {
...
-  name                = "loganalytics"
+resource "azurerm_application_insights" "example2" {
+  name                = "appinsights2"
+  location            = var.location
+  resource_group_name = azurerm_resource_group.example.name
+  application_type    = "web"
+}
+
+# Example: Alerting Action with result count trigger
+# Alert if more than three HTTP requests returned a >= 500 result code
```
Comment (mbfrahry) [inline website/docs/r/monitor_scheduled_query_rules_alert.html.markdown]: I don't think this is going to look quite right on the website. I'd throw a shorter description above the query attribute to explain what's going on

File: azurerm/helpers/azure/monitor.go [hunk 14]
```diff
@@ -2,8 +2,124 @@ package azure
...
+	if input != nil {
+		if input.ActionGroup != nil {
+			v["action_group"] = *input.ActionGroup
+		}
+		v["email_subject"] = input.EmailSubject
+		v["custom_webhook_payload"] = input.CustomWebhookPayload
+	}
+	result = append(result, v)
+
+	return result
```
[the remark this answers is not in the document]
Comment (mcdafydd) [reply azurerm/helpers/azure/monitor.go]: Thanks!

File: website/docs/r/monitor_scheduled_query_rules_alert.html.markdown (hunk 13 above)
Comment (mcdafydd) [reply website/docs/r/monitor_scheduled_query_rules_alert.html.markdown]: Let me know if this latest changeset is more aligned with your thoughts.  It's looking a lot nicer to me.

**Comment by mcdafydd** [comment]
@mbfrahry Ready for another peek.

**Comment by mbfrahry** [comment]
Thanks for this @mcdafydd! We'll get it into the next release!

**Pull request merged**